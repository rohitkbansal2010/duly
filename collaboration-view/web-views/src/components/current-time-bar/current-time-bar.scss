@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

$timeWidth: 6.25rem;

.barWrapper {
  position: absolute;
  z-index: 20;
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 0.13rem;
}

.currentTimeBar {
  display: flex;
  align-items: center;
  width: calc(100% - #{$timeWidth});
  height: 100%;
  background-color: $medium-red-violet;

  &:before {
    border-right: 0.4rem solid transparent;
    border-bottom: 1rem solid $medium-red-violet;
    border-left: 0.4rem solid transparent;
    border-radius: 0.1rem;
    transform: rotate(90deg);
    content: "";
  }
}

.currentTime {
  width: 100%;
  max-width: $timeWidth;
  color: $medium-red-violet;
  font-size: 1rem;
  line-height: 1.5rem;
  @include nexa-extra-bold-highlight;
}
