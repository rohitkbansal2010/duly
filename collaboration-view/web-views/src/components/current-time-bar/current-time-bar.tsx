import { useState } from 'react';
import { useInterval } from 'usehooks-ts';

import { OF_MILLISECONDS, SECONDS_PER_MINUTE } from '@constants';
import { calculateCurrentTimeBarOptions } from '@utils';

import styles from './current-time-bar.scss';

const workDayStart = window?.env?.WORK_DAY_START || '';
const workDayEnd = window?.env?.WORK_DAY_END || '';

export const CurrentTimeBar = () => {
  const now = new Date();
  const currentHour = now.getHours();

  const isBarVisible = !(currentHour < +workDayStart || currentHour >= +workDayEnd);

  const [ timerDuration, setTimerDuration ] = useState(
    (SECONDS_PER_MINUTE - now.getSeconds()) * OF_MILLISECONDS
  );

  const [ timeBarOptions, setTimeBarOptions ] = useState(calculateCurrentTimeBarOptions());

  useInterval(() => {
    setTimerDuration(SECONDS_PER_MINUTE * OF_MILLISECONDS);
    setTimeBarOptions(calculateCurrentTimeBarOptions());
  }, timerDuration);

  return isBarVisible ? (
    <div
      data-testid="current-time-bar__container"
      className={styles.barWrapper}
      style={{ top: `${timeBarOptions.position - 0.05}rem` }}
    >
      <span
        data-testid="current-time-bar__text"
        className={styles.currentTime}
      >
        {timeBarOptions.time}
      </span>
      <div className={styles.currentTimeBar} />
    </div>
  ) : null;
};
