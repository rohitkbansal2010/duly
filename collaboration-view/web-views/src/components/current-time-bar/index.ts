export { CurrentTimeBar } from './current-time-bar';
