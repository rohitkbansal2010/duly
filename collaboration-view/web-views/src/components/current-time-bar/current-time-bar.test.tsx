import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';

import { CurrentTimeBar } from '@components/current-time-bar/current-time-bar';

jest.spyOn(Date.prototype, 'getTimezoneOffset')
  .mockImplementation(() =>
    0);

describe('current-time-bar', () => {
  let Component: ReturnType<typeof mount>;
  const date = '2021-12-05T08:00:00';

  beforeAll(() => {
    jest.useFakeTimers('modern');
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  beforeEach(() => {
    jest.setSystemTime(new Date(date));
    Component = mount(<CurrentTimeBar />);
  });
  
  it('should render correctly', () => {
    expect(toJson(Component)).toMatchSnapshot();
  });

  it('should update time', () => {
    expect(Component.find({ 'data-testid': 'current-time-bar__text' }).text()).toBe('8:00 am');
    const oneHourInMiliseconds = 3600000;

    act(() => {
      jest.advanceTimersByTime(oneHourInMiliseconds);
      Component.update();
    });

    expect(Component.find({ 'data-testid': 'current-time-bar__text' }).text()).toBe('9:00 am');
  });

  it('should not render timebar if current hour is not in working span', () => {
    const newDate = '2021-12-05T20:00:00';
    jest.setSystemTime(new Date(newDate));

    const wrapper = mount(<CurrentTimeBar />);


    expect(wrapper.find({ 'data-testid': 'current-time-bar__container' }).exists()).toBe(false);
  });
});
