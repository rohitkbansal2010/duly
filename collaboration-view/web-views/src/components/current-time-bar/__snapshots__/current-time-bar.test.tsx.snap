// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`current-time-bar should render correctly 1`] = `
<CurrentTimeBar>
  <div
    className="barWrapper"
    data-testid="current-time-bar__container"
    style={
      Object {
        "top": "16.575rem",
      }
    }
  >
    <span
      className="currentTime"
      data-testid="current-time-bar__text"
    >
      8:00 am
    </span>
    <div
      className="currentTimeBar"
    />
  </div>
</CurrentTimeBar>
`;
